src/invariant/InvariantManager.h

```diff
@@ -60,13 +61,36 @@ class InvariantManager
         UnorderedMap<LedgerKey, LedgerEntry> const& restoredFromArchive,
         UnorderedMap<LedgerKey, LedgerEntry> const& restoredFromLiveState) = 0;
 
+    // This is used for expensive invariants that can't run in a blocking
+    // fashion, such as invariants that require scanning the entire BucketList.
+    // The invariant will periodically run on a background thread against the
+    // given ledger state snapshot. These invariants will only run if
+    // INVARIANT_EXTRA_CHECKS is enabled.
+    virtual void
+    runStateSnapshotInvariant(CompleteConstLedgerStatePtr ledgerState,
+                              InMemorySorobanState const& inMemorySnapshot) = 0;
+
     virtual void registerInvariant(std::shared_ptr<Invariant> invariant) = 0;
 
     virtual void enableInvariant(std::string const& name) = 0;
 
-    virtual void start(Application& app) = 0;
+    virtual void start(LedgerManager const& ledgerManager) = 0;
+
+    virtual bool shouldRunInvariantSnapshot() const = 0;
+
+    // Copy InMemorySorobanState for invariant checking. This is the only
+    // method that can access the private copy constructor of
+    // InMemorySorobanState.
+    virtual std::shared_ptr<InMemorySorobanState const>
+    copyInMemorySorobanStateForInvariant(
```

Comment:
this method seems a little out of place - since LM is managing the in-memory state, it should probably be the one copying it (we can probably use the member variable directly instead of a virtual method)